### ISSUE TITLE
Repeated calls to patch_all() eventually seem to infinite loop

### ISSUE THREAD

**Issue by palfrey** [DataDog/dd-trace-py#11297]
In ddtrace 2.14.4, if you call `ddtrace.patch_all()` many times, nothing bad seems to happen. In `2.15.1`, at some point, something seems to get stuck in an infinite loop. This occurred for us because part of the app setup in our FastAPI app did `patch_all()` and so pytest called it many times.

A workaround is something like the following

```python
_PATCH_ALL_DONE = False

def setup_datadog() -> None:
    global _PATCH_ALL_DONE
    if _PATCH_ALL_DONE is True:
        return
    ddtrace.patch_all()
    _PATCH_ALL_DONE = True
```

**Comment by palfrey** [comment]
> [@palfrey](https://github.com/palfrey) , are you using our `pytest` integration (eg: `pytest --ddtrace`) and using the `--ddtrace-patch-all` flag?

Nope. But having read https://ddtrace.readthedocs.io/en/stable/integrations.html#pytest I don't think that's what we want. We call `patch_all` in our app code and we want to test the code that does that, not call `patch_all` outside of our code.

**Comment by romainkomorndatadog** [comment]
@palfrey , thanks for the update, I think what you're doing makes sense.

@mabdinur , I don't think this is a CI app issue. It seems more like an issue with `ddtrace.patch_all()` not being idempotent.

**Comment by mabdinur** [comment]
@palfrey I wasn't able to reproduce this error using the following steps:

1. Install ddtrace==2.15.1, pytest==8.3.3, fastapi==0.115.5
2. Run the test below(ex: `pytest test_file.py -k test_patch_all`):

```
import ddtrace

def test_patch_all():
    for _ in range(10):
        ddtrace.patch_all()

        import fastapi

        app = fastapi.FastAPI()
```

Can you modify the test above to reproduce the error you are seeing in your testsuite?

**Comment by palfrey** [comment]
>     for _ in range(10):

Up to about 20 seems fine. At 21, 3.7s runtime; 22, 7.2s; 23, 13.72s, and I presume something in the range of doubling as that goes on.

**Comment by mabdinur** [comment]
## Investigation

Thanks for bringing this to our attention. [Patching ASM Common Modules](https://github.com/DataDog/dd-trace-py/blame/v2.15.1/ddtrace/_monkey.py#L238) appears to be increasing the duration of `ddtrace.patch_all()`. This PR may have introduced the performance regression in v2.15: https://github.com/DataDog/dd-trace-py/commit/4feddceb23c92d6091479a8fddd35e8eea4be8fa#diff-19592480afa34d43276c07c13325236a0adb62ef044a88b48bffce01992e6689. It appears that we are applying monkey patching to `BytesIO.read` and `StringIO.read` every time patch_all is called. This could be adding exponential overhead to all operations that involve reading strings. We should only apply patching once.

ddtrace patching for Exploit Prevention appears to be enabled by [default](https://github.com/DataDog/dd-trace-py/blob/main/ddtrace/_monkey.py#L237) (even if IAST/ASM are disabled). I will speak with the team about this design decision. 

## Workaround 

For now,  if your are not using the IAST or ASM products in CI, you can safely disable patching for exploit prevention by setting `DD_APPSEC_RASP_ENABLED=False`.

cc: @gnufede

**Comment by mabdinur** [comment]
Resolved by: https://github.com/DataDog/dd-trace-py/pull/11522

**Comment by mabdinur** [comment]
Fix was released in [v2.17.2](https://github.com/DataDog/dd-trace-py/releases/tag/v2.17.2)
